engine/access/rest/converter.go

```diff
@@ -247,21 +480,232 @@ func collectionGuaranteeResponse(flowCollGuarantee *flow.CollectionGuarantee) ge
 	return generated.CollectionGuarantee{
 		CollectionId: flowCollGuarantee.CollectionID.String(),
 		SignerIds:    signerIDs,
-		Signature:    base64.StdEncoding.EncodeToString(flowCollGuarantee.Signature.Bytes()),
+		Signature:    toBase64(flowCollGuarantee.Signature.Bytes()),
 	}
 }
 
-func blockSealsResponse(flowSeals []*flow.Seal) []generated.BlockSeal {
+func blockSealsResponse(flowSeals []*flow.Seal) ([]generated.BlockSeal, error) {
 	seals := make([]generated.BlockSeal, len(flowSeals))
 	for i, seal := range flowSeals {
-		seals[i] = blockSealResponse(seal)
+		sealResp, err := blockSealResponse(seal)
+		if err != nil {
+			return nil, err
+		}
+		seals[i] = sealResp
+	}
+	return seals, nil
+}
+
+func aggregatedApprovalSignaturesResponse(signatures []flow.AggregatedSignature) []generated.AggregatedSignature {
+	response := make([]generated.AggregatedSignature, len(signatures))
+	for i, signature := range signatures {
+
+		verifierSignatures := make([]string, len(signature.VerifierSignatures))
+		for y, verifierSignature := range signature.VerifierSignatures {
+			verifierSignatures[y] = toBase64(verifierSignature.Bytes())
+		}
+
+		signerIDs := make([]string, len(signature.SignerIDs))
+		for x, signerID := range signature.SignerIDs {
+			signerIDs[x] = signerID.String()
+		}
+
+		response[i] = generated.AggregatedSignature{
+			VerifierSignatures: verifierSignatures,
+			SignerIds:          signerIDs,
+		}
 	}
-	return seals
+
+	return response
 }
 
-func blockSealResponse(flowSeal *flow.Seal) generated.BlockSeal {
+func blockSealResponse(flowSeal *flow.Seal) (generated.BlockSeal, error) {
+	finalState := ""
+	if len(flowSeal.FinalState) > 0 { // todo(sideninja) this is always true?
+		finalStateBytes, err := flowSeal.FinalState.MarshalJSON()
+		if err != nil {
+			return generated.BlockSeal{}, err
+		}
+		finalState = string(finalStateBytes)
+	}
+
 	return generated.BlockSeal{
-		BlockId:  flowSeal.BlockID.String(),
-		ResultId: flowSeal.ResultID.String(),
+		BlockId:                      flowSeal.BlockID.String(),
+		ResultId:                     flowSeal.ResultID.String(),
+		FinalState:                   finalState,
+		AggregatedApprovalSignatures: aggregatedApprovalSignaturesResponse(flowSeal.AggregatedApprovalSigs),
+	}, nil
+}
+
+func collectionResponse(
+	collection *flow.LightCollection,
+	txs []*flow.TransactionBody,
+	link LinkGenerator,
+	expand map[string]bool,
+) (generated.Collection, error) {
+
+	expandable := &generated.CollectionExpandable{}
+
+	self, err := selfLink(collection.ID(), link.CollectionLink)
+	if err != nil {
+		return generated.Collection{}, err
+	}
+
+	var transactions []generated.Transaction
+	if expand[transactionsExpandable] {
+		for _, t := range txs {
```

Comment:
I believe we shouldn't, because it might not be expanded. I learned that from you 😉 😄  https://github.com/onflow/flow-go/pull/1666/files/4a05ced30ba727768f1ee38c2a43de4fc0f59edb#r765268409

Reply:
I just mean similar to the else case, where we allocate transactions within the conditional, but before the loop. If it's going to be a small list it probably doesn't matter
```golang
	var transactions []generated.Transaction
	if expand[transactionsExpandable] {
		transactions = make([]generated.Transaction, len(txs))
		for i, t := range txs {
			transactions[i] = *transactionResponse(t, nil, link)
		}
```